pico/configdriven/services/src/main/java/io/helidon/pico/configdriven/services/AbstractConfiguredServiceProvider.java

```diff
@@ -0,0 +1,832 @@
+import io.helidon.pico.DefaultServiceInfo;
+import io.helidon.pico.DefaultServiceInfoCriteria;
+import io.helidon.pico.Event;
+import io.helidon.pico.InjectionException;
+import io.helidon.pico.InjectionPointInfo;
+import io.helidon.pico.InjectionPointProvider;
+import io.helidon.pico.Phase;
+import io.helidon.pico.PicoException;
+import io.helidon.pico.PicoServiceProviderException;
+import io.helidon.pico.PicoServices;
+import io.helidon.pico.QualifierAndValue;
+import io.helidon.pico.ServiceInfo;
+import io.helidon.pico.ServiceInfoCriteria;
+import io.helidon.pico.ServiceProvider;
+import io.helidon.pico.ServiceProviderBindable;
+import io.helidon.pico.ServiceProviderProvider;
+import io.helidon.pico.configdriven.ConfiguredBy;
+import io.helidon.pico.services.AbstractServiceProvider;
+import io.helidon.pico.spi.InjectionResolver;
+
+import static io.helidon.pico.CallingContext.toErrorMessage;
+import static io.helidon.pico.configdriven.services.Utils.hasValue;
+import static io.helidon.pico.configdriven.services.Utils.isBlank;
+
+/**
+ * Abstract base for any config-driven-service.
+ *
+ * @param <T> the type of the service this provider manages
+ * @param <CB> the type of config beans that this service is configured by
+ */
+// special note: many of these methods are referenced in code generated code!
+public abstract class AbstractConfiguredServiceProvider<T, CB> extends AbstractServiceProvider<T>
+        implements ConfiguredServiceProvider<T, CB>,
+                   ServiceProviderProvider,
+                   InjectionPointProvider<T>,
+                   InjectionResolver {
+    private static final QualifierAndValue EMPTY_CONFIGURED_BY = DefaultQualifierAndValue.create(ConfiguredBy.class);
+    private static final CBInstanceComparator BEAN_INSTANCE_ID_COMPARATOR = new CBInstanceComparator();
+    private static final System.Logger LOGGER = System.getLogger(AbstractConfiguredServiceProvider.class.getName());
+    private static final LazyValue<InternalConfigBeanRegistry> CONFIG_BEAN_REGISTRY
+            = LazyValue.create(AbstractConfiguredServiceProvider::resolveConfigBeanRegistry);
+
+    private final AtomicReference<Boolean> isRootProvider = new AtomicReference<>();
+    private final AtomicReference<ConfiguredServiceProvider<T, CB>> rootProvider = new AtomicReference<>();
+    private final AtomicBoolean initialized = new AtomicBoolean();
+    private final AtomicReference<CallingContext> initializationCallingContext
+            = new AtomicReference<>();  // used only when we are in pico.debug mode
+    private final Map<String, Object> configBeanMap
+            = new TreeMap<>(BEAN_INSTANCE_ID_COMPARATOR);
+    private final Map<String, Optional<AbstractConfiguredServiceProvider<T, CB>>> managedConfiguredServicesMap
+            = new ConcurrentHashMap<>();
+
+    /**
+     * The default constructor.
+     */
+    protected AbstractConfiguredServiceProvider() {
+    }
+
+    @Override
+    protected System.Logger logger() {
```

Comment:
Methods have wrong order (public should be first, then protected, package local, private)